FileStatus.Unix/Process.Unix: align caching of user identity.

tmds opened the pull request:
The effective user, and group of the process won't change.

@dotnet/area-system-io ptal.

stephentoub commented:
> The effective user, and group of the process won't change.

What if the process calls setuid or seteuid?

tmds commented:
> What if the process calls setuid, seteuid, or setegid?

It is not supposed to be changed. In case someone changed it, we'd still be checking against the cached values.

In cases where you want to do something as a different user, it's very recommended to start a new process.

We already do this type of caching in Process.Unix:

https://github.com/dotnet/runtime/blob/de00deb924ebfc4418837002b60a1d31734b7e8d/src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.Unix.cs#L1066-L1068

carlossanlop reviewed (commented):
> In cases where you want to do something as a different user, it's very recommended to start a new process.

Can you recommend a link I can read explaining this expectation for Unix permissions?

> We could, if it makes a difference, avoid caching for root (egid = 0), since that is the most expected privileged user (which can use these calls).

Do you mean that if the initial process owner is root, we should always invoke `Interop.Sys.GetEUid()`/`Interop.Sys.GetEGid()` instead of returning the cached value, and if the value changes to a non-root user at some point, we then cache it, and start returning that? That would contradict the above statement, wouldn't it? That the recommendation is to start a new process if the user wants to switch to a different user.

tmds committed:
FileStatus.Unix/Process.Unix: align implementation.

Process: remove the user identity caching and extend the logic
to avoid retrieving the identity in most cases by checking
if all x-bits are set or not set.

FileStatus: use same group check as Process.

FileStatus: cache the read only flag instead of caching the
identity.

adamsitnik commented on src/libraries/Common/src/Interop/Unix/System.Native/Interop.IsMemberOfGroup.cs:
```diff
@@ -8,7 +8,26 @@ internal static partial class Interop
+                return true;
+            }
+
+            uint[]? groups = GetGroups();
```
Is `GetGroups()` used by any other methods? If not, we could either manually inline it or move the `Array.IndexOf` to it so we not need to allocate an array for the check.

adamsitnik commented on src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.Unix.cs:
```diff
@@ -769,29 +766,43 @@ private static bool IsExecutable(string fullPath)
+            uint euid = Interop.Sys.GetEUid();
+            if (euid == 0)
+            {
+                // We're root.
+                return true;
```
a very minor nit: we can put such short comments in the same line to save some space

```suggestion
                return true; // We're root.
```

adamsitnik commented on src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.Unix.cs:
```diff
@@ -1063,14 +1074,6 @@ private static unsafe void EnsureInitialized()
 
-                    s_euid = Interop.Sys.GetEUid();
-                    s_egid = Interop.Sys.GetEGid();
-                    s_groups = Interop.Sys.GetGroups();
```
I really like this change as it removes 3 sys-calls from the initialization logic that is used for every process, no matter if it's needed or not. 👍

adamsitnik reviewed (approved):
LGTM! It's great to see the identity caching being unified everywhere and the checks optimized for common cases.

adamsitnik commented:
@tmds could you please solve the conflicts and answer https://github.com/dotnet/runtime/pull/60160#discussion_r749272614 ?

tmds committed:
PR feedback
